### PR TITLE
Allow specification of number of columns with manual timestamps

### PR DESCRIPTION
**Pull request by rEtSaMfF**
Problem: 
I want to specify manual timestamps but I do not want to specify the full grid value.

Currently:
`vcsi.py -m 5:00,10:00,15:00 foo.mkv`
produces a 1x3 grid.

`vcsi.py -g 2x2 -m 5:00,10:00,15:00 foo.mkv`
produces a 1x3 grid.

`vcsi.py -g 4x1 -m 5:00,10:00,15:00 foo.mkv`
produces a 1x3 grid.

This patch allows specification of the number of columns.
`vcsi.py -g 2x2 -m 5:00,10:00,15:00 foo.mkv`
produces a 2x2 grid with one empty slot.

If we want this could easily be a `-c/--columns` flag.

I do want to implement an `--interval` or `--range` to avoid providing all of the manual timestamps on the command line.
Overall this helps facilitate the output when we do not know the `num_selected`

### CODE REVIEW

File: vcsi/vcsi.py [hunk 1]
```diff
@@ -948,6 +948,7 @@ def mxn_type(string):
     try:
         split = string.split("x")
         m = int(split[0])
+        assert (m > 0)
```
Comment (amietn) [inline vcsi/vcsi.py]: Raising an `argparse.ArgumentTypeError` sounds more appropriate in this case. Also since this makes sure that `m` is positive it would be better to check that `n` is also positive.

File: vcsi/vcsi.py [hunk 2]
```diff
@@ -1080,7 +1081,7 @@ def main():
         help="display frames on a mxn grid (for example 4x5)",
         dest="grid",
         type=mxn_type,
-        default=DEFAULT_GRID_SIZE)
+        default=None)
```
Comment (amietn) [inline vcsi/vcsi.py]: What is the reason for changing the default grid size value here?

File: vcsi/vcsi.py [hunk 3]
```diff
@@ -1325,8 +1326,30 @@ def process_file(path, args):
         args.metadata_horizontal_margin = args.metadata_margin
         args.metadata_vertical_margin = args.metadata_margin
 
+    if not args.grid:
```
Comment (amietn) [inline vcsi/vcsi.py]: Do you really need to know whether the default grid size value is used here?

File: vcsi/vcsi.py [hunk 4]
```diff
@@ -1325,8 +1326,30 @@ def process_file(path, args):
...
+    if args.manual_timestamps is not None:
+        mframes_size = len(args.manual_timestamps)
+        grid_size = args.grid.x * args.grid.y
+
+        args.num_selected = mframes_size
+        args.num_samples = mframes_size
+
+        if not mframes_size == grid_size:
+            # specified number of columns
+            y = math.ceil(mframes_size / args.grid.x)
```
Comment (amietn) [inline vcsi/vcsi.py]: Looks good

File: vcsi/vcsi.py [hunk 5]
```diff
@@ -1325,8 +1326,30 @@ def process_file(path, args):
...
+            args.grid = Grid(args.grid.x, y)
+
+    # interval frame selection
+
+    if args.num_selected < 1:
```
Comment (amietn) [inline vcsi/vcsi.py]: Since there is no `--interval` option at this point, leave this out of this pull request.

File: vcsi/vcsi.py (hunk 1 above)
Comment (rEtSaMfF) [inline vcsi/vcsi.py]: 1. the `assert` is caught by the try/except. We could specifically test and then raise the `argparse.ArgumentTypeError`. Or specify which exceptions are caught: `IndexError, ValueError, AssertionError`
2. My use case does not care about the `n`. As in I am okay with `-g 2x-1` or `-g 2x0`. This works better as a `-c/--columns` if that makes more sense.

File: vcsi/vcsi.py (hunk 2 above)
Comment (rEtSaMfF) [inline vcsi/vcsi.py]: Hmm, I do not remember, revert this.

File: vcsi/vcsi.py (hunk 3 above)
Comment (rEtSaMfF) [inline vcsi/vcsi.py]: I'll write a test if I can recall the case where this was needed.

File: vcsi/vcsi.py (hunk 5 above)
Comment (rEtSaMfF) [inline vcsi/vcsi.py]: Great, go ahead and make whatever changes you need.

I will put in some tests for the edge cases.

**Comment by rEtSaMfF** [comment]
Great, go ahead and make whatever changes you need.

I will put in some tests for the edge cases.

File: vcsi/vcsi.py (hunk 1 above)
Comment (amietn) [inline vcsi/vcsi.py]: I like option 1. here. Catching different exception types and displaying appropriate error message in each case.

File: vcsi/vcsi.py (hunk 2 above)
Comment (amietn) [inline vcsi/vcsi.py]: If you can't remember then it's probably not needed.

File: vcsi/vcsi.py (hunk 3 above)
Comment (amietn) [inline vcsi/vcsi.py]: Please don't write a test for code that not needed, simply remove  it.

File: tests/test_input.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    assert_raises(ArgumentTypeError, mxn_type, '4xa')
+
+    assert_raises(ArgumentTypeError, mxn_type, '4x4.1')
+
+    test_grid = mxn_type('4x-1')
+
+    assert_equals(test_grid.x, 4)
+    assert_equals(test_grid.y, -1)
+
+    assert_raises(ArgumentTypeError, mxn_type, '4x1x4')
```
Comment (rEtSaMfF) [inline tests/test_input.py]: This produces `Grid(x=4, y=1)`

Is this our desired behavior?

**Comment by rEtSaMfF** [comment]
Changes are here with some unit tests.

Please comment on the ambiguous test case.

And let me know if you want the whole thing squashed.

File: tests/test_input.py (hunk 6 above)
Comment (amietn) [inline tests/test_input.py]: It is not :) Good catch!

The form should definitely be `mxn` and only one "x" should appear. In this case we should display an error message.

Also I think it only makes sense to have strictly positive values for the grid size. We should display an error if either `grid.x` or `grid.y` is <= 0

Would you like to fix that?

**Comment by amietn** [comment]
Thanks! This is looking good. I replied to your comment about the test case over there. The whole thing will be squashed when I merge it, no worries.

**Comment by amietn** [comment]
Merged.
